src/fabric/test/fabric2_index_tests.erl

```diff
@@ -0,0 +1,302 @@
+            ])
+        }
+    }.
+
+
+index_process_cleanup_test_() ->
+    {
+        "Test fabric process cleanup in indexing module",
+        {
+            foreach,
+            fun setup/0,
+            fun cleanup/1,
+            [
+                ?TDEF_FE(updater_processes_start),
+                ?TDEF_FE(updater_processes_stop),
+                ?TDEF_FE(indexing_can_be_disabled),
+                ?TDEF_FE(handle_indexer_blowing_up)
+            ]
+        }
+    }.
+
+
+setup() ->
+    meck:new(config, [passthrough]),
+    meck:expect(config, get_integer, fun
+        ("fabric", "index_updater_delay_msec", _) -> 200;
+        ("fabric", "index_updater_resolution_msec", _) -> 100;
+
+        (_, _, Default) -> Default
+    end),
+    meck:expect(config, get_boolean, fun
+        ("fabric", "index_updater_enabled", _) -> true;
+        (_, _, Default) -> Default
+    end),
+
+    Indices = application:get_env(fabric, indices, []),
+
+    Ctx = test_util:start_couch([fabric]),
+
+    % Db1 has a valid design doc, a deleted one and one with "autoupdate":false
+    {ok, Db1} = fabric2_db:create(?tempdb(), [?ADMIN_CTX]),
+    {_, _} = create_doc(Db1, <<"_design/doc1">>),
+
+    DDocId2 = <<"_design/doc2">>,
+    {DDocId2, {Pos, Rev}} = create_doc(Db1, DDocId2),
+    Delete2 = #doc{id = DDocId2, revs = {Pos, [Rev]}, deleted = true},
+    {ok, _} = fabric2_db:update_doc(Db1, Delete2),
+
+    NoAutoUpdate = {[{<<"autoupdate">>, false}]},
+    {_, _} = create_doc(Db1, <<"_design/doc3">>, NoAutoUpdate),
+
+    % Db2 doesn't have any desig documents
+    {ok, Db2} = fabric2_db:create(?tempdb(), [?ADMIN_CTX]),
+
+    #{db1 => Db1, db2 => Db2, ctx => Ctx, indices => Indices}.
+
+
+cleanup(#{db1 := Db1, db2 := Db2, ctx := Ctx, indices := Indices}) ->
+    catch fabric2_db:delete(fabric2_db:name(Db1), []),
+    catch fabric2_db:delete(fabric2_db:name(Db2), []),
+
+    test_util:stop_couch(Ctx),
+    application:set_env(fabric, indices, Indices),
+
+    meck:unload().
+
+
+register_index_works(_) ->
+    reset_callbacks(),
+
+    Mod1 = fabric2_test_callback1,
+    fabric2_index:register_index(Mod1),
+    Indices1 = application:get_env(fabric, indices, []),
+    ?assertEqual([Mod1], Indices1),
+
+    Mod2 = fabric2_test_callback2,
+    fabric2_index:register_index(Mod2),
+    Indices2 = application:get_env(fabric, indices, []),
+    ?assertEqual(lists:sort([Mod1, Mod2]), lists:sort(Indices2)).
+
+
+single_update(#{db1 := Db}) ->
+    reset_callbacks(),
+
+    Mod = fabric2_test_callback3,
+    setup_callback(Mod),
+    create_doc(Db),
+
+    meck:wait(Mod, build_indices, 2, 2000),
+    ?assertEqual(1, meck:num_calls(Mod, build_indices, 2)).
+
+
+multiple_updates(#{db1 := Db}) ->
+    reset_callbacks(),
+
+    Mod = fabric2_test_callback4,
+    setup_callback(Mod),
+    create_docs(Db, 10),
+
+    % should be called at least once
+    meck:wait(Mod, build_indices, 2, 2000),
+
+    % Maybe called another time or two at most
```

Comment:
I'm not seeing why your assertion of `=< 3` is correct. Is it because of the jitter and frequency checks and the fact you're only waiting 500ms?

Reply:
1st call is call from the pervious test line

2nd is if that might be called while we are inserting the docs.

3rd is the one called after we are done inserting all the docs.